internal/transport/grpchttp2/framer.go

```diff
@@ -0,0 +1,246 @@
+type FrameHeader struct {
+	// Size is the size of the frame's payload without the 9 header bytes.
+	// As per the HTTP/2 spec, size can be up to 3 bytes, but only frames
+	// up to 16KB can be processed without agreement.
+	Size uint32
+	// Type is a byte that represents the Frame Type.
+	Type FrameType
+	// Flags is a byte representing the flags set on this Frame.
+	Flags Flags
+	// StreamID is the ID for the stream which this frame is for. If the
+	// frame is connection specific instead of stream specific, the
+	// streamID is 0.
+	StreamID uint32
+}
+
+// Frame represents an HTTP/2 Frame.
+type Frame interface {
+	Header() *FrameHeader
+}
+
+type DataFrame struct {
+	hdr  *FrameHeader
+	free func()
+	Data []byte
+}
+
+func (f *DataFrame) Header() *FrameHeader {
+	return f.hdr
+}
+
+func (f *DataFrame) Free() {
+	if f.free != nil {
+		f.free()
+	}
+}
+
+type HeadersFrame struct {
+	hdr      *FrameHeader
+	free     func()
+	HdrBlock []byte
+}
+
+func (f *HeadersFrame) Header() *FrameHeader {
+	return f.hdr
+}
+
+func (f *HeadersFrame) Free() {
+	if f.free != nil {
+		f.free()
+	}
+}
+
+type RSTStreamFrame struct {
+	hdr  *FrameHeader
+	Code ErrCode
+}
+
+func (f *RSTStreamFrame) Header() *FrameHeader {
+	return f.hdr
+}
+
+type SettingsFrame struct {
+	hdr      *FrameHeader
+	free     func()
+	settings []Setting
+}
+
+func (f *SettingsFrame) Header() *FrameHeader {
+	return f.hdr
+}
+
+type PingFrame struct {
+	hdr  *FrameHeader
+	free func()
+	Data []byte
+}
+
+func (f *PingFrame) Header() *FrameHeader {
+	return f.hdr
+}
+
+func (f *PingFrame) Free() {
+	if f.free != nil {
+		f.free()
+	}
+}
+
+type GoAwayFrame struct {
+	hdr          *FrameHeader
+	free         func()
+	LastStreamID uint32
+	Code         ErrCode
+	DebugData    []byte
+}
+
+func (f *GoAwayFrame) Header() *FrameHeader {
+	return f.hdr
+}
+
+func (f *GoAwayFrame) Free() {
+	if f.free != nil {
+		f.free()
+	}
+}
+
+type WindowUpdateFrame struct {
+	hdr *FrameHeader
+	Inc uint32
+}
+
+func (f *WindowUpdateFrame) Header() *FrameHeader {
+	return f.hdr
+}
+
+type ContinuationFrame struct {
+	hdr      *FrameHeader
+	free     func()
+	HdrBlock []byte
+}
+
+func (f *ContinuationFrame) Header() *FrameHeader {
+	return f.hdr
+}
+
+func (f *ContinuationFrame) Free() {
+	if f.free != nil {
+		f.free()
+	}
+}
+
+// MetaHeadersFrame is not a Frame Type that appears on the HTTP/2 Spec. It is
+// a representation of the merging and decoding of all the Headers and
+// Continuation frames on a Stream.
+type MetaHeadersFrame struct {
+	hdr    *FrameHeader
+	Fields []hpack.HeaderField
+}
+
+func (f *MetaHeadersFrame) Header() *FrameHeader {
+	return f.hdr
+}
+
+// Framer represents a Framer used in gRPC-Go.
+type Framer interface {
+	// SetMetaDecoder will set a decoder for the framer. When the decoder is
+	// set, ReadFrame will parse the header values, merging all Headers and
+	// Continuation frames.
+	SetMetaDecoder(d *hpack.Decoder)
```

Comment:
What is the behavior when a decoder is not set for a framer? Is there a default one?

Reply:
Yes, it will just return a single `HeadersFrame` or `ContinuationFrame`, with the raw bytes instead of the decoded data. That is the way the `net/x` framer works. We could make the decoding the default behaviour and remove the (current) default one, since gRPC has no need for the raw bytes, we just take a look at the decoded headers. What are your thoughts?

Reply:
If we stick with having both behaviours, do you think I should put the defualt behaviour in the doc for the `SetMetaDecoder` or the `ReadFrame` method?

Reply:
Is there value in this decoder being configurable? If not, we can make decoding the default behavior.